README.md
# @sinoui/theme [![npm version](https://img.shields.io/npm/v/@sinoui/theme)](https://www.npmjs.com/package/@sinoui/theme) [![downloads](https://img.shields.io/npm/dm/@sinoui/theme)](https://www.npmjs.com/package/@sinoui/theme)

用来创建 sinoui 的主题对象的库。[文档](https://sinoui.github.io/theme/)

## 安装

```shell
yarn add @sinoui/theme
```

## 本地开发

项目中有以下有用的命令。

### `yarn start`

在开发和监听模式下启动项目。当代码发生变化时就会重新编译代码。它同时会实时地向你汇报项目中的代码错误。

### `yarn build`

打包，并将打包文件放在`dist`文件夹中。使用 rollup 对代码做优化并打包成多种格式（`Common JS`，`UMD`和`ES Module`）。

### `yarn lint`

`yarn lint`会检查整个项目是否有代码错误、风格错误。

开启 vscode 的 eslint、prettier 插件，在使用 vscode 编码时，就会自动修正风格错误、提示语法错误。

### `yarn format`

`yarn format`可以自动调整整个项目的代码风格问题。

### `yarn test`

`yarn test`以监听模式启动 jest，运行单元测试。

开启 vscode 的 jest 插件，会在文件变化时自动运行单元测试。

### 预览文档

```shell
yarn doc:dev
```

### 编译并打包文档

```shell
yarn doc:publish
```

### 发布文档

在发布文档之前，在`package.json`中配置好`homepage`，如下所示：

```json
{
  "homepage": "https://sinouiincubator.github.io/editable-data-table"
}
```

配置完之后就可以执行下面的命令行发布文档：

```shell
yarn doc:publish
```


CHANGELOG.md
# 版本变更说明

## v1.0.0-beta.8 2020.3.18

- fix(typography): 修复 button 字体字号不正确的 bug

## v1.0.0-beta.7 2020.3.5

- fix(typography): 修复字体样式对象的类型声明不兼容 styled-components 的缺陷
- fix(shadows): 修复海拔为 2 和 3 的阴影样式不正确的错误

## v1.0.0-beta.5 2020.2.28

- fix: 修复 DefaultTheme 类型错误

## v1.0.0-beta.3 2020.2.25

- fix: 修复打包失败的缺陷

## v1.0.0-beta.1 2020.2.24

- feat(typography): 采用新的 Material Design 标准设置文字排版分类
- feat(typography): 简化自定义文字排版的方式
- feat(typography): `typography` 新增 `pxToRem` 函数，用来转换 px 到 rem
- feat(typography): 可以定制文字排版分类样式
- feat: 新增 ThemeStyle 组件，用来定义全局 css 样式
- chore(colors): 拆分 colors.ts 为一个颜色一个 ts 文件

### 破坏性变更

#### 创建主题的方式

之前的方式是：

```ts
import {
  createTheme,
  createPalette,
  createTypography,
  colors,
} from '@sinoui/theme';

const theme = createTheme({
  typography: createTypography({
    fontSize: 12,
  }),
  palette: createPalette({
    primary: colors.red,
  }),
});
```

为了简化创建主题，提供了新的方式是：

```ts
import { createTheme } from '@sinoui/theme';
import red from '@sinoui/theme/colors/red';

const theme = createTheme({
  typography: {
    fontSize: 12,
  },
  palette: {
    primary: colors.red,
  },
});
```

虽然仍然支持旧的方式，但是不再推荐使用。

#### 文字排版

文字排版分类发生变化，之前的分类是：

- dispay4
- display3
- display2
- display1
- subheading
- title
- headline
- body2
- body1
- caption
- button

新的文字排版分类：

- h1
- h2
- h3
- h4
- h5
- h6
- subtitle1
- subtitle2
- body1
- body2
- caption
- button

#### 颜色模块 colors module

1. colors 模块剔除了非标准的颜色，包括：

   - transparent
   - fullBlack
   - darkBlack
   - lightBlack
   - minBlack
   - faintBlack
   - fullWhite
   - darkWhite
   - lightWhite

2. colors 模块将 white 和 black 迁移到 colors/common 中

   之前的方式：

   ```ts
   import { colors } from '@sinoui/theme';

   console.log(colors.white, colors.black);
   ```

   新的方式：

   ```ts
   import common from '@sinoui/theme/colors';

   console.log(common.white, common.black);
   ```

3. colors 模块去掉自定义颜色系列

   - colorBlack
   - colorDeepGray
   - colorBrown
   - colorGray
   - colorKelly
   - colorLigntYellow

4. colors 模块去掉颜色系列中的 `w1`, `w2`, `w3`

#### 调色板

调色板发生了重大变化，去掉了很多可以通过计算轻而易举得到的颜色值。新版本的调色板，推荐大家在日常开发中多使用[polished](https://polished.js.org/)来扩展颜色。

新的调色板使用方式：

- 调色板不为单独的组件定义颜色，组件内部颜色在组件内部定义或者生成
- 调色板提供了 `palette.grey`，组件可以使用此灰色集合来定义内部颜色
- 使用 [polished](https://polished.js.org/) 提供的方法扩展颜色

去掉的颜色值：

- 去掉了整个 `danger` 颜色分类，使用 `warnning` 代替
- 去掉了整个 `input`

  - helperText, labelText --> palette.text.secondary
  - disabled --> palette.text.disabled
  - inputText --> palette.text.primary
  - bottomLine --> 在组件中直接定义即可，没必要放在主题定制中

- 去掉部分 `text` 中的属性：

  - icon --> sinoui-components 未使用
  - divider --> palette.divider
  - lightDivider --> sinoui-components 未使用
  - snackbar --> palette.getContrastText(snackbar 组件的背景色)

- 去掉了大部分 `background` 中的属性，只保留 default 和 paper：

  - appBar --> palette.grey[100]
  - contentFrame --> palette.grey[200]
  - divider --> palette.divider（见后面的说明）
  - chips --> palette.grey[300]
  - chipsImg --> palette.grey[400]
  - chipsBut --> palette.grey[400]
  - chipsButHover --> palette.grey[500]
  - tooltip --> grey[600]
  - transparent --> 直接使用 `transparent` 关键字
  - sendcondaryTransparent --> palette.text.primary
  - snackbar --> palette.common.black

  对于 sinoui-components 中使用的 `divider`，表示的是选中后的背景色，推荐使用 `palette.action.selected` 代替。如果觉得颜色不合适，也可以使用灰色：`palette.grey[200]`。

- `palette.black`, `palette.white` 分别变更为 `palette.common.black`，`palette.common.white`

- 去掉了整个 `shades`

  sinoui-components 使用 `shades` 来取文本颜色：`palette.shades.dark.text.primary`。替换成新的获取文本颜色方式：`palette.getContrastText(背景色)`。

另一个重大变更：**不再提供颜色分类的集合，替换成包含 main, dark, light, contrastText 的对象**。这些颜色分类包括：`primary`, `secondary`, `error`, `success`, `info`, `success`。

旧的颜色分类集合：

```ts
{
  primary: {
    50: '#xxx',
    100: '#xxx',
    200: '#xxx',
    //...
  },
  secondary: {
    // ...
  },
  //...
}
```

新的颜色对象：

```ts
{
  primary: {
    main: '#xxx',
    light: '#xxx',
    dark: '#xxx',
    contrastText: '#xxx',
  },
  secondary: {
    // ...
  },
  // ...
}
```

`primary.main` 对应此颜色集合的主颜色，`primary.light` 是颜色集合的浅色变体，`primary.dark` 是颜色集合的深色变体，`primary.contrastText` 是在主颜色背景上的文字颜色。

如果需要获取更深或者更浅的颜色，可以使用 [polished](https://polished.js.org/) 的 `darken` 和 `lighten` 方法来获取，如：

```ts
import { darken, lighten } from 'polished';

const light = lighten(0.2, palette.primary.main);
const dark = darken(0.2, palette.primary.main);
```

#### zIndex

去掉了组件库中不使用的 z-index：

- drawerOverlay - 遮罩层
- navDrawer - 导航区域
- dialogOverlay - 对话框遮罩层
- layer - 新的应用层
- menu - 菜单
- mobileStepper

## v0.3.4 - 2019.12.19

- fix: 修复在 IE 中无法使用的缺陷

## v0.3.3 - 2019.10.22

- chore: 升级 polished@3.4.1

## v0.3.2 - 2019.10.22

- chore: polished 降版本到 1.9.2

## v0.3.1 - 2019.10.22

- chore: 补充缺失的导出方法

## v0.3.0 - 2019.10.22

- feat: 断点 ( #2 )


src/types.ts
/* eslint-disable @typescript-eslint/no-empty-interface */

/**
 * 色系
 */
export interface ColorSeries {
  50: string;
  100: string;
  200: string;
  300: string;
  400: string;
  500: string;
  600: string;
  700: string;
  800: string;
  900: string;
  A100: string;
  A200: string;
  A400: string;
  A700: string;
}
/* --------- 调色板 ------- */

export type ShadeKey = keyof ColorSeries;

/**
 * 调色板颜色
 */
export interface PaletteColor {
  /**
   * 主颜色
   */
  main: string;
  /**
   * 深色
   */
  dark: string;
  /**
   * 浅色
   */
  light: string;
  /**
   * 对比文本的颜色
   */
  contrastText: string;
}

/**
 * 文本颜色
 */
export interface TypeText {
  /**
   * 重要文本的颜色
   */
  primary: string;
  /**
   * 次要文本的颜色，如
   */
  secondary: string;
  /**
   * 不可用文本颜色。
   */
  disabled: string;
  /**
   * hint 模式下的文本颜色。
   */
  hint: string;
}

/**
 * 按钮颜色
 */
export interface TypeAction {
  /**
   * 可用按钮的文本颜色
   */
  active: string;
  /**
   * 鼠标移动到按钮上时的背景色
   */
  hover: string;
  /**
   * hover状态下的透明度
   */
  hoverOpacity: number;
  /**
   * 按钮被选中时的背景色
   */
  selected: string;
  /**
   * 选中状态下的透明度
   */
  selectedOpacity: number;
  /**
   * 按钮不可用时的文本颜色
   */
  disabled: string;
  /**
   * 按钮不用时的背景色
   */
  disabledBackground: string;
}

/**
 * 背景颜色
 */
export interface TypeBackground {
  /**
   * 默认背景颜色
   */
  default: string;
  /**
   * 纸张背景颜色
   */
  paper: string;
}

/**
 * 类型颜色
 */
export interface TypeObject {
  /**
   * 文本颜色
   */
  text: TypeText;
  /**
   * 按钮颜色
   */
  action: TypeAction;
  /**
   * 背景颜色
   */
  background: TypeBackground;
  /**
   * 分割线颜色
   */
  divider: string;
}

type PaletteType = 'light' | 'dark';

export interface Palette extends TypeObject {
  type: PaletteType;
  /**
   * 品牌色，主颜色。
   */
  primary: PaletteColor;
  /**
   * 辅助色。
   */
  secondary: PaletteColor;
  /**
   * 错误状态颜色。
   */
  error: PaletteColor;
  /**
   * 警告状态颜色。
   */
  warning: PaletteColor;
  /**
   * 普通信息颜色。
   */
  info: PaletteColor;
  /**
   * 成功状态颜色。
   */
  success: PaletteColor;
  /**
   * 灰色色系
   */
  grey: ColorSeries;
  /**
   * 获取在指定背景颜色上的符合对比度标准的文本颜色（即 Material Design 中的 `on color`）
   *
   * @param color 背景颜色
   */
  getContrastText(color: string): string;
  /**
   * 通用颜色
   */
  common: { white: string; black: string };
}

export type PaletteColorOption =
  | ColorSeries
  | {
      /**
       * 主颜色
       */
      main: string;
      /**
       * 深色
       */
      dark?: string;
      /**
       * 浅色
       */
      light?: string;
      /**
       * 对比文本的颜色
       */
      contrastText?: string;
    };

export interface PaletteOptions {
  /**
   * 主颜色
   */
  primary?: PaletteColorOption;
  /**
   * 辅助色。默认为主题色。
   */
  secondary?: PaletteColorOption;
  /**
   * 错误状态颜色。
   */
  error?: PaletteColorOption;
  /**
   * 警告状态颜色。
   */
  warning?: PaletteColorOption;
  /**
   * 普通信息颜色。
   */
  info?: PaletteColorOption;
  /**
   * 成功状态颜色。
   */
  success?: PaletteColorOption;
  /**
   * 灰色色系
   */
  grey?: Partial<ColorSeries>;
  /**
   * 颜色模式。
   */
  type?: PaletteType;
  /**
   * 设置文本颜色。
   */
  text?: Partial<TypeText>;
  /**
   * 设置按钮颜色。
   */
  action?: Partial<TypeAction>;
  /**
   * 设置背景颜色。
   */
  background?: Partial<TypeBackground>;
  /**
   * 分割线颜色
   */
  divider?: string;
}

/* ---------------typography----------------*/

/**
 * 字重
 */
export type FontWeight = 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900;

/**
 * 字体配置项
 */
export type TypographyConstants = Partial<ThemeTypography>;

/**
 * 文本主题
 */
export interface ThemeTypography {
  /**
   * 字体家族
   */
  fontFamily: string;
  /**
   * 极细字重
   */
  fontWeightLight: FontWeight;
  /**
   * 正常字重
   */
  fontWeightRegular: FontWeight;
  /**
   * 中等字重
   */
  fontWeightMedium: FontWeight;
  /**
   * 粗体字重
   */
  fontWeightBold: FontWeight;
  /**
   * 控制字体大小的参考值。默认为 `14px`。
   */
  fontSize: number;
  /**
   * `<html>`标签的字体大小，即浏览器默认字体大小。默认为 `16px`。
   */
  htmlFontSize: number;
  /**
   * 一级标题样式
   */
  h1: React.CSSProperties;
  /**
   * 二级标题样式
   */
  h2: React.CSSProperties;
  /**
   * 三级标题样式
   */
  h3: React.CSSProperties;
  /**
   * 四级标题样式
   */
  h4: React.CSSProperties;
  /**
   * 五级标题样式
   */
  h5: React.CSSProperties;
  /**
   * 六级标题样式
   */
  h6: React.CSSProperties;
  /**
   * 副标题1样式
   */
  subtitle1: React.CSSProperties;
  /**
   * 副标题2样式
   */
  subtitle2: React.CSSProperties;
  /**
   * 主体文本1样式
   */
  body1: React.CSSProperties;
  /**
   * 主体文本2样式
   */
  body2: React.CSSProperties;
  /**
   * 按钮文本样式
   */
  button: React.CSSProperties;
  /**
   * 说明文本样式
   */
  caption: React.CSSProperties;
  /**
   * overline 样式
   */
  overline: React.CSSProperties;

  /**
   * 将 px 值转换为 rem 值
   */
  pxToRem: (px: number) => string;
}

/**
 * 阴影主题
 */
export interface ThemeShadows {
  [index: number]: string;
}

/**
 * 变换配置项
 */
export interface TransitionOptions {
  duration?: number;
  easing?: string;
  delay?: number;
}

/**
 * 缓动函数
 */
export interface ThemeTransitionsEasing {
  easeInOut: string;
  easeOut: string;
  easeIn: string;
  sharp: string;
}

/**
 * css3变换时长
 */
export interface ThemeTransitionsDuration {
  shortest: number;
  shorter: number;
  short: number;
  standard: number;
  complex: number;
  enteringScreen: number;
  leavingScreen: number;
}

/**
 * css3变换主题
 */
export interface ThemeTransitions {
  easing: ThemeTransitionsEasing;
  duration: ThemeTransitionsDuration;
  create: (props: string | string[], options?: TransitionOptions) => string;
  getAutoHeightDuration: (height: number) => number;
}

/**
 * zIndex主题
 */
export interface ThemeZIndex {
  mobileStepper: number;
  menu: number;
  appBar: number;
  drawerOverlay: number;
  navDrawer: number;
  modal: number;
  dialogOverlay: number;
  dialog: number;
  layer: number;
  popover: number;
  snackbar: number;
  tooltip: number;
}

/**
 * 留白主题
 */
export interface ThemeSpacing {
  unit: number;
}

/**
 * 主题配置项
 */
export interface ThemeOptions {
  palette?: Palette;
  typography?: TypographyConstants;
  shadows?: ThemeShadows;
}

interface Breakpoints {
  /**
   * 超小尺寸断点。手机。
   */
  xs: number;
  /**
   * 小尺寸断点。平板。
   */
  sm: number;
  /**
   * 中等尺寸断点。笔记本电脑。
   */
  md: number;
  /**
   * 大尺寸断点。台式机电脑屏幕。
   */
  lg: number;
  /**
   * 超大尺寸断点。
   */
  xl: number;
  [other: string]: number;
}

/**
 * 主题
 */
export interface Theme {
  /**
   * 调色板
   */
  palette: Palette;
  /**
   * 字体
   */
  typography: ThemeTypography;
  /**
   * 阴影
   */
  shadows: ThemeShadows;
  /**
   * 过度动画
   */
  transitions: ThemeTransitions;
  /**
   * 层叠
   */
  zIndex: ThemeZIndex;
  /**
   * 间距
   */
  spacing: ThemeSpacing;
  /**
   * 断点
   */
  breakpoints: Breakpoints;
}

declare module 'styled-components' {
  export interface DefaultTheme extends Theme {}
}
